service/worker/deployment/deployment_workflow.go

```diff
@@ -190,45 +197,151 @@ func (d *DeploymentWorkflowRunner) handleRegisterWorker(ctx workflow.Context, ar
+	// use lock to enforce only one update at a time
+	err := d.lock.Lock(ctx)
+	if err != nil {
+		d.logger.Error("Could not acquire workflow lock")
+		return nil, serviceerror.NewDeadlineExceeded("Could not acquire workflow lock")
+	}
+	d.pendingUpdates++
+	defer func() {
+		d.pendingUpdates--
+		d.lock.Unlock()
+	}()
+
+	// wait until series workflow started
+	err = workflow.Await(ctx, func() bool { return d.State.StartedSeriesWorkflow })
+	if err != nil {
+		d.logger.Error("Update canceled before series workflow started")
+		return nil, serviceerror.NewDeadlineExceeded("Update canceled before series workflow started")
+	}
+
+	// apply changes to "current"
+	if set := args.SetCurrent; set != nil {
+		if set.LastBecameCurrentTime == nil {
+			d.State.IsCurrent = false
+		} else {
+			d.State.IsCurrent = true
+			d.State.LastBecameCurrentTime = set.LastBecameCurrentTime
+		}
+		d.updateMemo(ctx)
+
+		// sync to task queues
+		syncReq := &deploymentspb.SyncUserDataRequest{
+			Deployment: d.State.WorkerDeployment,
+		}
+		for tqName, byType := range d.State.TaskQueueFamilies {
+			for tqType, data := range byType.TaskQueues {
+				syncReq.Sync = append(syncReq.Sync, &deploymentspb.SyncUserDataRequest_SyncUserData{
+					Name: tqName,
+					Type: enumspb.TaskQueueType(tqType),
+					Data: d.dataWithTime(data),
+				})
+			}
+		}
+		activityCtx := workflow.WithActivityOptions(ctx, defaultActivityOptions)
+		err = workflow.ExecuteActivity(activityCtx, d.a.SyncUserData, syncReq).Get(ctx, nil)
+		if err != nil {
+			// TODO: if this fails, should we roll back anything?
+			return nil, err
+		}
+	}
+
+	// apply changes to metadata
+	for key, payload := range args.UpdateMetadata.GetUpsertEntries() {
+		d.State.Metadata[key] = payload
+	}
+	for _, key := range args.UpdateMetadata.GetRemoveEntries() {
+		delete(d.State.Metadata, key)
+	}
+
+	return &deploymentspb.SyncDeploymentStateResponse{
+		DeploymentLocalState: d.State,
+	}, nil
+}
+
+func (d *DeploymentWorkflowRunner) dataWithTime(data *deploymentspb.TaskQueueData) *deploymentspb.TaskQueueData {
+	data = common.CloneProto(data)
+	data.LastBecameCurrentTime = d.State.LastBecameCurrentTime
+	return data
+}
+
 func (d *DeploymentWorkflowRunner) handleDescribeQuery() (*deploymentspb.QueryDescribeDeploymentResponse, error) {
 	return &deploymentspb.QueryDescribeDeploymentResponse{
-		DeploymentLocalState: d.DeploymentLocalState,
+		DeploymentLocalState: d.State,
 	}, nil
 }
 
-/*
 // updateMemo should be called whenever the workflow updates it's local state: "is_current_deployment"
-func (d *DeploymentWorkflowRunner) updateMemo() {
+func (d *DeploymentWorkflowRunner) updateMemo(ctx workflow.Context) {
+	workflow.UpsertMemo(ctx, map[string]any{
+		DeploymentMemoField: &deploymentspb.DeploymentWorkflowMemo{
+			Deployment:          d.State.WorkerDeployment,
```

Comment:
Nit: to make it consistent with everywhere else
```suggestion
			Deployment:          d.State.Deployment,
```